Yank/repex.py

```diff
@@ -2099,6 +2215,65 @@ def _attempt_swap(self, replica_i, replica_j):
             self._n_accepted_matrix[thermodynamic_state_i, thermodynamic_state_j] += 1
             self._n_accepted_matrix[thermodynamic_state_j, thermodynamic_state_i] += 1
 
+    # -------------------------------------------------------------------------
+    # Internal-usage: Online Analysis.
+    # -------------------------------------------------------------------------
+
+    @mpi.on_single_node(rank=0, broadcast_result=False, sync_nodes=False)
+    @mpi.delayed_termination
+    @mmtools.utils.with_timer('Computing online free energy estimate')
+    def _run_online_analysis(self):
+        # Prep the reporter
+        self._reporter.close()
+        self._reporter.open('r')
+        if self._last_mbar_f_k is None:
+            # Backwards search for last weights, don't need the 0th iteration because the weights are 0
+            try:
+                for index in range(self._iteration, 0, -1):
+                    f_k = self._reporter.read_mbar_weights(index)
+                    # Find an f_k that is not all zeros
+                    if not np.all(f_k == 0):
+                        self._last_mbar_f_k = f_k
+                        break  # Don't need to continue the loop if we already found one
+            except (IndexError, KeyError):  # No such f_k written yet (or variable created), no need to do anything
+                pass
+        # Start the analysis
+        bump_error_counter = False
+        analysis = RepexPhase(self._reporter, analysis_kwargs={'initial_f_k': self._last_mbar_f_k})
+        # Indices for online analysis, "i'th index, j'th index"
+        idx, jdx = 0, -1
+        timer = mmtools.utils.Timer()
+        timer.start("MBAR")
+        logger.debug("Computing free energy with MBAR...")
+        try:
+            mbar = analysis.mbar
+            free_energy, err_free_energy = analysis.get_free_energy()
+            output_fe = free_energy[idx, jdx]
+            output_dfe = err_free_energy[idx, jdx]
+            logger.debug("Current Free Energy Estimate is {} +- {} kT".format(output_fe, output_dfe))
+            self._last_mbar_f_k = mbar.f_k
+            if np.isnan(output_dfe):  # Trap a case when errors don't converge (usually due to under sampling)
+                output_dfe = np.inf
+        except ParameterError as e:
```

Comment:
With poorly overlapping data that is also poorly sampled, MBAR can throw odd numerical issues when it tries to solve itself. There are 2 main ones I know of, the case where the W_kn sums dont add to 1, throwing the parameter error, and the case where the error estimate does not converge and throws either a nan or an inf. The only times I really see this is in very, very poorly overlapped data, which may happen.

Because poor overlapping data (due to poorly sampled data) should not be a crash criteria, I trap it a few times to see if it clears up after a few iterations. I don't let it run on forever though and eventually crash it, reporting the stack of errors it banked along the way.